src/py/flwr/common/auth_plugin/keycloak_plugin.py

```diff
@@ -0,0 +1,271 @@
+                    (
+                        _REFRESH_TOKEN_HEADER,
+                        refresh_token,
+                    ),
+                )
+            )
+            return True
+
+        return False
+
+
+class KeycloakUserPlugin(UserAuthPlugin):
+    """Abstract Flower User Auth Plugin class."""
+
+    def __init__(self, config: dict[str, Any], config_path: Path):
+        self.access_token = config["access-token"]
+        self.refresh_token = config["refresh-token"]
+        self.config = config
+        self.config_path = config_path
+
+    @staticmethod
+    def login(
+        login_details: dict[str, str],
+        config: dict[str, Any],
+        federation: str,
+        exec_stub: ExecStub,
+    ) -> dict[str, Any]:
+        """Read relevant auth details from federation config."""
+        timeout = int(login_details.get("expires_in", "600"))
+        interval = int(login_details.get("interval", "5"))
+        device_code = login_details.get("device_code")
+        verification_uri_complete = login_details.get("verification_uri_complete")
+
+        if device_code is None or verification_uri_complete is None:
+            typer.secho(
+                "❌ Missing information for Keycloak login .",
+                fg=typer.colors.RED,
+                bold=True,
+            )
+            raise typer.Exit(code=1)
+
+        typer.secho(
+            "Please login with your user credentials here: "
+            f"{verification_uri_complete}",
+            fg=typer.colors.BLUE,
+        )
+        start_time = time.time()
+        time.sleep(interval)
+
+        while (time.time() - start_time) < timeout:
+            auth_details = {"device_code": device_code}
+            res: GetAuthTokenResponse = exec_stub.GetAuthToken(
+                GetAuthTokenRequest(auth_details=auth_details)
+            )
+
+            access_token = res.auth_tokens.get("access_token")
+            refresh_token = res.auth_tokens.get("refresh_token")
+
+            if access_token and refresh_token:
+                config = {}
+                config["auth-type"] = "keycloak"
+                config["access-token"] = access_token
+                config["refresh-token"] = refresh_token
+
+                typer.secho(
+                    "Login successful. You can now execute an authenticated "
+                    "`flwr run` to the SuperLink.",
+                    fg=typer.colors.GREEN,
+                    bold=False,
+                )
+                return config
+
+            time.sleep(interval)
+
+        typer.secho(
+            "❌ Timeout, failed to sign in.",
+            fg=typer.colors.RED,
+            bold=True,
+        )
+        raise typer.Exit(code=1)
+
+    def provide_auth_details(self, metadata: Metadata) -> Metadata:
+        """Provide relevant auth tokens in the metadata."""
+        metadata.append(
+            (
+                _AUTH_TOKEN_HEADER,
+                self.access_token.encode("utf-8"),
```

Comment:
May I know why we use UTF-8 encoding here for the tokens? What do tokens look like?